File: 2022/haskell/day02/solve.hs
{-# LANGUAGE OverloadedStrings #-}

import Data.Char (ord)
import Data.List (sort)

parseInput :: String -> [(Int, Int)]
parseInput input = [(ord elf - ord 'A', ord me - ord 'X') | (elf : _ : me : _) <- lines input]

solvePart1 :: String -> Int
solvePart1 = sum . map score . parseInput

solvePart2 :: String -> Int
solvePart2 = sum . map selectscore . parseInput

score :: (Int, Int) -> Int
score (elf, me) = (me - elf + 1) `mod` 3 * 3 + me + 1

selectscore :: (Int, Int) -> Int
selectscore (elf, me) = score (elf, (elf + me - 1) `mod` 3)

main :: IO ()
main =
  do
    input <- readFile "input.txt"
    putStrLn $ "Part one: " ++ show (solvePart1 input)
    putStrLn $ "Part two: " ++ show (solvePart2 input)

File: 2022/python/day02/solve.py
input = open("input.txt").read()
invs = [(ord(item[0]) - ord('A'), ord(item[2]) - ord('X'))
        for item in input.splitlines()]


def score(elf, me):
    if elf == me:
        return 3 + me + 1
    elif elf == (me + 1) % 3:
        return 0 + me + 1
    return 6 + me + 1


def selectscore(elf, result):
    return score(elf, (elf + result) % 3)


print("Part 1:", sum([score(elf, me) for (elf, me) in invs]))
print("Part 2:", sum([selectscore(elf, me - 1) for (elf, me) in invs]))
